File: packages/dart/test/fixtures_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:test/test.dart';
import 'package:kreuzberg_cloud_sdk/kreuzberg_cloud_sdk.dart';

ExtractionResult _loadFixture(String name) {
  // spec/fixtures lives at the repo root; from packages/dart/ the path is ../../spec/fixtures/
  final file = File('../../spec/fixtures/$name');
  final json = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  return ExtractionResult.fromJson(json);
}

void main() {
  group('ExtractionResult wire shape', () {
    group('minimal payload', () {
      test('should_deserialize_content_and_mime_type', () {
        final result = _loadFixture('extraction_result_minimal.json');
        expect(result.content, 'Hello');
        expect(result.mimeType, 'text/plain');
      });

      test('should_deserialize_empty_tables_list', () {
        final result = _loadFixture('extraction_result_minimal.json');
        expect(result.tables, isEmpty);
      });

      test('should_have_non_null_metadata', () {
        final result = _loadFixture('extraction_result_minimal.json');
        expect(result.metadata, isNotNull);
        expect(result.metadata.format, isNull);
      });
    });

    group('PDF fixture', () {
      test('should_deserialize_content_and_mime_type', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.mimeType, 'application/pdf');
        expect(result.content, contains('quarterly financial results'));
      });

      test('should_deserialize_quality_score', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.qualityScore, closeTo(0.91, 0.001));
      });

      test('should_have_three_pages', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.pages, isNotNull);
        expect(result.pages, hasLength(3));
        expect(result.pages!.first.pageNumber, 1);
      });

      test('table_cells_should_be_nested_string_array', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.tables, hasLength(1));
        expect(result.tables.first.cells, [
          ['Header A', 'Header B'],
          ['Row 1A', 'Row 1B'],
        ]);
      });

      test('table_markdown_should_match_fixture', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(
          result.tables.first.markdown,
          '| Header A | Header B |\n|----------|----------|\n| Row 1A   | Row 1B   |',
        );
        expect(result.tables.first.pageNumber, 1);
      });

      test('metadata_title_and_authors_should_match_fixture', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.metadata.title, 'Q1 2024 Financial Report');
        expect(result.metadata.authors, ['Alice Müller', 'Bob Chen']);
      });

      test('format_discriminator_should_resolve_to_ocr_variant', () {
        final result = _loadFixture('extraction_result_pdf.json');
        expect(result.metadata.format, isNotNull);
        expect(result.metadata.format, isA<FormatMetadataOcr>());
      });

      test('ocr_variant_fields_should_match_fixture', () {
        final result = _loadFixture('extraction_result_pdf.json');
        final ocr = result.metadata.format! as FormatMetadataOcr;
        expect(ocr.language, 'eng');
        expect(ocr.outputFormat, 'markdown');
        expect(ocr.psm, 3);
        expect(ocr.tableCount, 1);
      });

      test('format_map_dispatch_should_reach_ocr_branch', () {
        final result = _loadFixture('extraction_result_pdf.json');
        final dispatched = result.metadata.format!.map(
          ocr: (ocr) => 'got-ocr:${ocr.language}',
          archive: (_) => 'archive',
          csv: (_) => 'csv',
          email: (_) => 'email',
          excel: (_) => 'excel',
          html: (_) => 'html',
          image: (_) => 'image',
          pptx: (_) => 'pptx',
          pst: (_) => 'pst',
          text: (_) => 'text',
          xml: (_) => 'xml',
        );
        expect(dispatched, 'got-ocr:eng');
      });
    });

    group('XLSX fixture with children', () {
      test('should_deserialize_content_and_mime_type', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(
          result.mimeType,
          'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        );
        expect(result.content, contains('Sheet1'));
      });

      test('should_have_three_tables', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(result.tables, hasLength(3));
      });

      test('first_table_cells_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(result.tables.first.cells, [
          ['Department', 'Budget', 'Actual'],
          ['Engineering', '50000', '48500'],
          ['Marketing', '30000', '31200'],
        ]);
      });

      test('metadata_title_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(result.metadata.title, 'Annual Budget 2024');
      });

      test('format_discriminator_should_resolve_to_excel_variant', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(result.metadata.format, isNotNull);
        expect(result.metadata.format, isA<FormatMetadataExcel>());
      });

      test('excel_variant_fields_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        final excel = result.metadata.format! as FormatMetadataExcel;
        expect(excel.sheetCount, 3);
        expect(excel.sheetNames, ['Q1', 'Q2', 'Q3']);
      });

      test('format_map_dispatch_should_reach_excel_branch', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        final dispatched = result.metadata.format!.map(
          excel: (excel) => 'got-excel:${excel.sheetCount}',
          archive: (_) => 'archive',
          csv: (_) => 'csv',
          email: (_) => 'email',
          html: (_) => 'html',
          image: (_) => 'image',
          ocr: (_) => 'ocr',
          pptx: (_) => 'pptx',
          pst: (_) => 'pst',
          text: (_) => 'text',
          xml: (_) => 'xml',
        );
        expect(dispatched, 'got-excel:3');
      });

      test('should_have_one_child_archive_entry', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        expect(result.children, isNotNull);
        expect(result.children, hasLength(1));
      });

      test('child_path_and_mime_type_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        final child = result.children!.first;
        expect(child.path, 'xl/embeddings/Doc1.docx');
        expect(
          child.mimeType,
          'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        );
      });

      test('child_nested_extraction_result_should_deserialize', () {
        final result = _loadFixture(
          'extraction_result_xlsx_with_children.json',
        );
        final childResult = result.children!.first.result;
        expect(childResult.content, contains('methodology notes'));
        expect(childResult.metadata.title, 'Methodology Notes');
        expect(childResult.metadata.authors, ['Alice Müller']);
        expect(childResult.tables, isEmpty);
      });
    });

    group('DOCX fixture with revisions', () {
      test('should_deserialize_content_and_mime_type', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        expect(
          result.mimeType,
          'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        );
        expect(result.content, contains('project timeline'));
      });

      test('should_have_three_revisions', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        expect(result.revisions, isNotNull);
        expect(result.revisions, hasLength(3));
      });

      test('first_revision_should_be_insertion', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final first = result.revisions!.first;
        expect(first.kind, RevisionKind.insertion);
        expect(first.revisionId, '42');
        expect(first.author, 'Alice');
        expect(first.timestamp, '2024-03-15T10:30:00Z');
      });

      test('second_revision_should_be_deletion', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final second = result.revisions![1];
        expect(second.kind, RevisionKind.deletion);
        expect(second.revisionId, '43');
        expect(second.author, 'Alice');
      });

      test('third_revision_should_be_comment', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final third = result.revisions![2];
        expect(third.kind, RevisionKind.comment);
        expect(third.revisionId, '44');
        expect(third.author, 'Bob');
        expect(third.timestamp, '2024-03-16T09:00:00Z');
      });

      test('revision_kind_enum_values_should_be_correct_type', () {
        expect(RevisionKind.insertion, isA<RevisionKind>());
        expect(RevisionKind.deletion, isA<RevisionKind>());
        expect(RevisionKind.comment, isA<RevisionKind>());
      });

      test('should_have_two_extracted_keywords', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        expect(result.extractedKeywords, isNotNull);
        expect(result.extractedKeywords, hasLength(2));
      });

      test('keywords_should_have_yake_algorithm_and_scores', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final keywords = result.extractedKeywords!;
        for (final keyword in keywords) {
          expect(keyword.algorithm, KeywordAlgorithm.yake);
          expect(keyword.score, greaterThan(0.0));
        }
      });

      test('first_keyword_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final first = result.extractedKeywords!.first;
        expect(first.text, 'project timeline');
        expect(first.score, closeTo(0.42, 0.001));
        expect(first.algorithm, KeywordAlgorithm.yake);
        expect(first.positions, [4, 30]);
      });

      test('second_keyword_should_match_fixture', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final second = result.extractedKeywords![1];
        expect(second.text, 'delivery date');
        expect(second.score, closeTo(0.38, 0.001));
        expect(second.positions, [55, 68]);
      });

      test('keyword_algorithm_toString_should_contain_yake', () {
        final result = _loadFixture(
          'extraction_result_docx_with_revisions.json',
        );
        final first = result.extractedKeywords!.first;
        expect(first.algorithm.toString(), contains('yake'));
      });
    });

    group('URI fixture', () {
      test('should_deserialize_content_and_mime_type', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        expect(result.mimeType, 'text/markdown');
        expect(result.content, contains('Research Notes'));
      });

      test('should_have_three_uris', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        expect(result.uris, isNotNull);
        expect(result.uris, hasLength(3));
      });

      test('all_uris_should_have_non_empty_url', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        for (final uri in result.uris!) {
          expect(uri.url, isNotEmpty);
        }
      });

      test('first_uri_should_be_hyperlink', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        final first = result.uris!.first;
        expect(first.kind, UriKind.hyperlink);
        expect(first.url, 'https://docs.example.com');
        expect(first.label, 'official documentation');
        expect(first.page, 1);
      });

      test('second_uri_should_be_email', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        final second = result.uris![1];
        expect(second.kind, UriKind.email);
        expect(second.url, 'mailto:support@example.com');
        expect(second.label, 'support@example.com');
        expect(second.page, 1);
      });

      test('third_uri_should_be_anchor', () {
        final result = _loadFixture('extraction_result_with_uris.json');
        final third = result.uris![2];
        expect(third.kind, UriKind.anchor);
        expect(third.url, '#fig-1');
        expect(third.label, 'Figure 1');
        expect(third.page, 1);
      });
    });

    group('round-trip deserialisation', () {
      for (final fixture in [
        'extraction_result_minimal.json',
        'extraction_result_pdf.json',
        'extraction_result_xlsx_with_children.json',
        'extraction_result_docx_with_revisions.json',
        'extraction_result_with_uris.json',
      ]) {
        test('should_deserialize_${fixture}_without_error', () {
          expect(() => _loadFixture(fixture), returnsNormally);
        });
      }
    });
  });
}

File: packages/typescript/tests/fixtures.test.ts
/**
 * Deserialization tests that lock in the ExtractionResult wire shape.
 *
 * Each test loads a JSON fixture from spec/fixtures/, casts it to the
 * generated TypeScript type, and asserts on specific field values. A
 * regression in the generated schema will make type-level assertions fail
 * at compile time and value-level assertions fail at runtime.
 */

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import type { components } from "../src/_generated/schema";

type ExtractionResult = components["schemas"]["ExtractionResult"];

const FIXTURES = resolve(__dirname, "../../../spec/fixtures");

function load(name: string): ExtractionResult {
	return JSON.parse(readFileSync(resolve(FIXTURES, name), "utf-8")) as ExtractionResult;
}

// ---------------------------------------------------------------------------
// Smoke: all fixtures must parse as valid JSON and cast without error
// ---------------------------------------------------------------------------

describe("all fixtures deserialize without error", () => {
	const fixtureNames = [
		"extraction_result_minimal.json",
		"extraction_result_pdf.json",
		"extraction_result_xlsx_with_children.json",
		"extraction_result_docx_with_revisions.json",
		"extraction_result_with_uris.json",
	] as const;

	it.each(fixtureNames)("parses %s", (name) => {
		expect(() => load(name)).not.toThrow();
	});
});

// ---------------------------------------------------------------------------
// Minimal fixture
// ---------------------------------------------------------------------------

describe("minimal ExtractionResult", () => {
	it("should_have_correct_content_and_mime_type", () => {
		const r = load("extraction_result_minimal.json");
		expect(r.content).toBe("Hello");
		expect(r.mime_type).toBe("text/plain");
	});

	it("should_have_empty_tables_array", () => {
		const r = load("extraction_result_minimal.json");
		expect(r.tables).toEqual([]);
	});

	it("should_have_metadata_object", () => {
		const r = load("extraction_result_minimal.json");
		expect(r.metadata).toBeDefined();
		expect(typeof r.metadata).toBe("object");
	});
});

// ---------------------------------------------------------------------------
// PDF fixture
// ---------------------------------------------------------------------------

describe("PDF ExtractionResult", () => {
	it("should_have_pdf_mime_type", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.mime_type).toBe("application/pdf");
	});

	it("should_have_authors_as_string_array", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.metadata.authors).toEqual(["Alice Müller", "Bob Chen"]);
	});

	it("should_have_quality_score_close_to_expected_value", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.quality_score).toBeCloseTo(0.91);
	});

	it("Table_cells_is_a_nested_string_array", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.tables[0]?.cells).toEqual([
			["Header A", "Header B"],
			["Row 1A", "Row 1B"],
		]);
	});

	it("should_have_table_page_number_and_markdown", () => {
		const r = load("extraction_result_pdf.json");
		const table = r.tables[0];
		expect(table?.page_number).toBe(1);
		expect(table?.markdown).toContain("Header A");
	});

	it("PDF_metadata_format_uses_ocr_format_type_discriminator", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.metadata.format?.format_type).toBe("ocr");
	});

	it("should_have_three_pages", () => {
		const r = load("extraction_result_pdf.json");
		expect(r.pages).toHaveLength(3);
		expect(r.pages?.[0]?.page_number).toBe(1);
	});
});

// ---------------------------------------------------------------------------
// XLSX with children fixture
// ---------------------------------------------------------------------------

describe("XLSX ExtractionResult with children", () => {
	it("should_have_xlsx_mime_type", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		expect(r.mime_type).toBe("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
	});

	it("should_have_excel_format_type_discriminator", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		expect(r.metadata.format?.format_type).toBe("excel");
	});

	it("should_have_one_child_archive_entry", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		expect(r.children).toHaveLength(1);
	});

	it("child_should_have_correct_path_and_mime_type", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		const child = r.children?.[0];
		expect(child?.path).toBe("xl/embeddings/Doc1.docx");
		expect(child?.mime_type).toBe("application/vnd.openxmlformats-officedocument.wordprocessingml.document");
	});

	it("child_carries_nested_ExtractionResult_with_content", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		const nested = r.children?.[0]?.result;
		expect(nested?.content).toBeTruthy();
		expect(nested?.tables).toEqual([]);
	});

	it("should_have_three_tables_across_sheets", () => {
		const r = load("extraction_result_xlsx_with_children.json");
		expect(r.tables).toHaveLength(3);
		expect(r.tables[0]?.cells[0]).toEqual(["Department", "Budget", "Actual"]);
	});
});

// ---------------------------------------------------------------------------
// DOCX with revisions fixture
// ---------------------------------------------------------------------------

describe("DOCX ExtractionResult with revisions", () => {
	it("should_have_three_revisions", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		expect(r.revisions).toHaveLength(3);
	});

	it("first_revision_is_insertion_by_alice", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		const rev = r.revisions?.[0];
		expect(rev?.kind).toBe("insertion");
		expect(rev?.author).toBe("Alice");
		expect(rev?.revision_id).toBe("42");
		expect(rev?.timestamp).toBe("2024-03-15T10:30:00Z");
	});

	it("revisions_cover_insertion_deletion_and_comment_kinds", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		const kinds = new Set(r.revisions?.map((rev) => rev.kind));
		expect(kinds).toEqual(new Set(["insertion", "deletion", "comment"]));
	});

	it("should_have_two_extracted_keywords", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		expect(r.extracted_keywords).toHaveLength(2);
	});

	it("first_keyword_has_yake_algorithm_and_positive_score", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		const kw = r.extracted_keywords?.[0];
		expect(kw?.algorithm).toBe("yake");
		expect(kw?.score).toBeGreaterThan(0);
		expect(kw?.text).toBe("project timeline");
	});

	it("keyword_positions_are_integer_arrays", () => {
		const r = load("extraction_result_docx_with_revisions.json");
		const kw = r.extracted_keywords?.[0];
		expect(kw?.positions).toEqual([4, 30]);
	});
});

// ---------------------------------------------------------------------------
// URI fixture
// ---------------------------------------------------------------------------

describe("ExtractionResult with URIs", () => {
	it("should_have_three_uris", () => {
		const r = load("extraction_result_with_uris.json");
		expect(r.uris).toHaveLength(3);
	});

	it("all_uris_have_non_empty_url", () => {
		const r = load("extraction_result_with_uris.json");
		const urls = r.uris?.map((u) => u.url) ?? [];
		expect(urls.every((url) => url.length > 0)).toBe(true);
	});

	it("uri_kinds_and_labels_match_fixture_values", () => {
		const r = load("extraction_result_with_uris.json");
		const uris = r.uris ?? [];
		expect(uris[0]?.kind).toBe("hyperlink");
		expect(uris[0]?.url).toBe("https://docs.example.com");
		expect(uris[0]?.label).toBe("official documentation");
		expect(uris[1]?.kind).toBe("email");
		expect(uris[2]?.kind).toBe("anchor");
	});

	it("all_uris_are_on_page_1", () => {
		const r = load("extraction_result_with_uris.json");
		const pages = r.uris?.map((u) => u.page) ?? [];
		expect(pages.every((p) => p === 1)).toBe(true);
	});
});
